### ISSUE TITLE
Real Time Recommendations (new users)

### ISSUE THREAD

**Issue by Emtschikay** [MLnick/elasticsearch-vector-scoring#8]
Hello,

first of all, thanks for the great work. I'm currently looking into it and got the demo working as posted on the IBM repository.

Going through it, I have one question. If there's a new user coming in and I get some implicit ratings from him, I would at a certain point want to make recommendations to her/him. Now, if (s)he hasn't been included in the training data, no user vector exists for him. Is there any way I can get recommendations for her/him without retraining the whole model?

Thanks! I hope this implementation sees an update to work with ES 5.5+ 😃

**Comment by MLnick** [comment]
Hi there

Thanks for the feedback. I hope to get to the ES 5.5 update soon!

The cold-start problem (where a new user or item comes into the system) is a complex and quite extensively studied area of recommenders.

As you mention, typically one has to re-train the model to get factors vectors for the new entity. (It is possible with factor models to do a form of real-time incremental, approximate model update but I haven't done so here yet).

The other ways to try to address cold-start usually focus on using some additional features such as item content, or richer models that include some form of user / item / event metadata that still allow prediction in the case of cold-start.

For the simple factor model approach, one way of doing dealing with it, is to treat an anonymous (or new) user as a "collection of item vectors" that represent the items the user has interacted with (e.g. the items they have viewed in a session). One can average these item vectors and use that simulated factor vector to compute item similarity and in that way have some form of personalized recommendation.

It is more tricky for new items as that will usually require a model update.

Hope that makes sense.

**Comment by Emtschikay** [comment]
Hey Nick,

thanks for answering. Yes, that makes perfect sense, actually I thought of the same thing as a quick solution - averaging or at least aggregating the past clicks in some way in the product space and looking for the nearest neighbour(s). Thanks a lot for digging into this.

One last remark: since I'm running a 5.4.2 cluster, I updated and built the plugin from source. Strangely enough, ElasticSearch was complaining when I tried to install it from local (something along the lines of "unknown plugin"). This went away when I uploaded it to github and installed it from there. Same also goes for the version you provide here - if I download it manually and try to install it from local, it doesn't work. Remote everything is perfect. Just to let you know.

Best
